### PR TITLE
fix: do not error on unknown os_family grain

### PR DESCRIPTION
**Pull request by Sxderp**
If the formula was ran on a system that reported an os_family that
wasn't one of "Debian", "RedHat", "Arch", "Suse" then the map.jinja
template would fail to render with "'NoneType' is not iterable.

This occurs because grains.filter_by will return None when it fails
match the grain to the input dictionary. The value is then blindly
passed into a dict.update() which causes the failure.

In this patch we ensure that the default values, as defined in
defaults.yaml, are always applied when grain matching fails.

### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [x] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Related issues and/or pull requests

None.

### Describe the changes you're proposing

Fix map.jinja template rendering. Do not fail when the system reports an unknown os_family grain.

### Pillar / config required to test the proposed changes

A system that reports a os_family value that is different from "Debian", "RedHat", "Arch", "Suse" . This error was brought to our attention when the formula was mistakenly ran on a MacOS machine. The formula isn't supposed to work on MacOS, but this error is platform agnostic. There may be some Distribution using firewalld that does not match the known os_family grains.

### Debug log showing how the proposed changes work

I have a traceback. The fix prevents the traceback. Should more debugging be required then I can get it.
```
local:
    Data failed to compile:
----------
    Rendering SLS 'cos-sdaniele3-dev:firewalld' failed: Jinja error: 'NoneType' object is not iterable
/var/cache/salt/minion/files/cos-sdaniele3-dev/firewalld/map.jinja(20):
---
[...]
    'Suse': {},
  }, grain='os_family', merge=salt['pillar.get']('firewalld:lookup'))
%}

{#- Merge the flavor_map to the default settings #}
{% do default_settings.firewalld.update(os_family_map) %}    <======================

{#- Merge in salt:lookup pillar #}
{% set firewalld = salt['pillar.get'](
    'firewalld',
    default=default_settings.firewalld,
[...]
---
Traceback (most recent call last):
  File "/opt/salt/lib/python2.7/site-packages/salt/utils/templates.py", line 393, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/opt/salt/lib/python2.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/opt/salt/lib/python2.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 6, in top-level template code
  File "/opt/salt/lib/python2.7/site-packages/jinja2/environment.py", line 1073, in make_module
    return TemplateModule(self, self.new_context(vars, shared, locals))
  File "/opt/salt/lib/python2.7/site-packages/jinja2/environment.py", line 1152, in __init__
    body_stream = list(template.root_render_func(context))
  File "/var/cache/salt/minion/files/cos-sdaniele3-dev/firewalld/map.jinja", line 20, in top-level template code
    {% do default_settings.firewalld.update(os_family_map) %}
TypeError: 'NoneType' object is not iterable
```

### Documentation checklist

- [x] Updated the `README` (e.g. `Available states`).
- [x] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### CODE REVIEW

**Comment by myii** [comment]
@Sxderp Thanks for providing this fix.  The method you're proposing won't cut it as-is.  Let me attempt a short explanation.

There's a long history of discussion about `map.jinja` throughout this organisation.  In summary, there are different levels of `map.jinja` maturity found across the formulas.  Unfortunately, the existing `map.jinja` here is closer to `v1` than anything else.  The current recommendation is to switch the version shown in the `template-formula`, which could be considered a `v3` (taking the best bits of all of the `v2` examples out there).  You'll find it's pretty sturdy all round.  That being said, we're already working towards a `v4` but that doesn't concern us right now for this fix.

What we _don't_ want to do is go off in a different direction, so that this becomes an even bigger change to make further down the line.  We also want to avoid breaking changes where possible.  Shifting to a new `default` root in `firewalld/defaults.yaml` is a step to be avoided.

Long story short, the right solution here is to refactor the current setup to introduce the `v3` style from the [`template-formula`](https://github.com/saltstack-formulas/template-formula/blob/93af398cda5c7f2289c3c7211722601d92d5e3e7/TEMPLATE/map.jinja).  The "blank" `os*map.yaml` files need to be copied across as well, with any minor adjustments made.

I could do this pretty quickly but I'll give you the chance to try it out if you fancy taking it on yourself.  Let me know how you'd like to proceed.

**Comment by Sxderp** [comment]
I'll work on it. But I do think I see a problem with the current version of the template. I should probably make this an issue in that repository:

From what I can see [assigning tplroot](https://github.com/saltstack-formulas/template-formula/blob/93af398cda5c7f2289c3c7211722601d92d5e3e7/TEMPLATE/map.jinja#L5) doesn't work if your formula is in a subdirectory. Example:

```
$ tree -F states/
states/
├── remote/
│   └── firewalld/
│       └── map.jinja
└── self-maintained/
    ├── linux-clevis/
    └── linux-tang/
```

Your `tplroot` is going to be "remote" is it not? You can't then assume the `tplroot` is the same as  the key name in the pillar (assuming the key name is the same as the folder is poor regardless, I think).

I haven't run the code so I'm not sure if `tpldir` populates the full path or not, but I assume it does.

**Comment by myii** [comment]
> I'll work on it. But I do think I see a problem with the current version of the template. I should probably make this an issue in that repository:
> 
> From what I can see [assigning tplroot](https://github.com/saltstack-formulas/template-formula/blob/93af398cda5c7f2289c3c7211722601d92d5e3e7/TEMPLATE/map.jinja#L5) doesn't work if your formula is in a subdirectory. Example:
> 
> ```
> $ tree -F states/
> states/
> ├── remote/
> │   └── firewalld/
> │       └── map.jinja
> └── self-maintained/
>     ├── linux-clevis/
>     └── linux-tang/
> ```

Yes, this has come up in discussions in the Salt repo:

* https://github.com/saltstack/salt/issues/41195#issuecomment-577973741
  - You can see the breakdown in the table there

What I suggest for this PR is to avoid `tplroot` and even `tpldir` (since we've had other issues with that in the past) and stick to literal values for now.  There is a potential solution but it will take time to finalise.

**Comment by Sxderp** [comment]
The formula now has the stub map files as demonstrated in the template formula.

While I included it in the patch I'm not a fan of using `config.get`. I feel is muddies the data mapping of formulas. User supplied data should be consumed from the pillar. With `config.get` the data could come from multiple different sources.

If `config.get` does data merging (not sure, haven't looked deep into it) then it could be even harder to track down "what's going on".

**Comment by myii** [comment]
> The formula now has the stub map files as demonstrated in the template formula.

Thanks, I'll get around to reviewing and merging this soon.

> While I included it in the patch I'm not a fan of using `config.get`. I feel it muddies the data mapping of formulas. User supplied data should be consumed from the pillar. With `config.get` the data could come from multiple different sources.

We have had _extensive_ discussions about this and on many different occasions.  Suffice to say that this is based on recommendations from SaltStack itself.  If you're on Slack or IRC, feel free to drop in to the #formulas channel (#saltstack-formulas on IRC) and it would be easier to cover there.

> If `config.get` does data merging (not sure, haven't looked deep into it) then it could be even harder to track down "what's going on".

I understand your concerns but having used it extensively for a while now, it's actually not as difficult as it may appear.

**Comment by myii** [comment]
@Sxderp Thanks for this PR, what you've provided is fine.  I made a few minor changes for other reasons:

1. When evaluating `map.jinja` changes, I am using InSpec to verify it, so I added the tests in #42 and then had to rebase this PR on top -- hence, the force push.
1. We consider `pillar.get` => `config.get` a `BREAKING CHANGE`, which I appended to the relevant commit message.
1. The map verification had to be updated due to the expected change introduced by this PR (i.e. the `arch` entry in the map).

I'll merge this once the tests complete.

**Comment by myii** [comment]
@Sxderp Merged.
